addons/payment_paypal/controllers/main.py

```diff
@@ -24,52 +26,104 @@ def paypal_dpn(self, **data):
+
+        The validation is done in four steps:
+          - 1: Make a POST request to Paypal with the `tx`, the GET param received with the PDT,
+               data, and the two other required params `cmd` and `at`.
+          - 2: PayPal sends back a response text starting with either 'SUCCESS' or 'FAIL'. If the
+               validation was a success, the notification data are appended to the response text as
+               a string formatted as follows: 'SUCCESS\nparam1=value1\nparam2=value2\n...'
+          - 3: Extract the notification data and process these instead of the PDT data.
+          - 4: Return an empty HTTP 200 response (done at the end of the route controller).
+        See https://developer.paypal.com/docs/api-basics/notifications/payment-data-transfer/.
+
+        :param dict data: The data whose authenticity to check
+        :return: The retrieved notification data
+        :raise: ValidationError if the authenticity could not be verified
+        """
+        def _parse_pdt_validation_response(_response_content):
+            """ Parse the validation response and return the parsed notification data.
+
+            :param str _response_content: The PDT validation request response
+            :return: The parsed notification data
+            :rtype: dict
+            """
+            _response_items = _response_content.rstrip('\n').split('\n')
+            _success = _response_items[0]
+            if _success == 'SUCCESS':
+                _notification_data = {}
+                for _notification_data_param in _response_items[1:]:
+                    _key, _raw_value = _notification_data_param.split('=', 1)
+                    _notification_data[_key] = urls.url_unquote_plus(_raw_value)
+                return _notification_data
+            return None
+
+        tx_sudo = request.env['payment.transaction'].sudo()._get_tx_from_feedback_data(
+            'paypal', data
+        )
+        if 'tx' not in data:  # We did not receive PDT data but directly notification data
+            # When PDT is not enabled, PayPal sends directly the notification data instead. We can't
+            # verify them but we can process them as is.
+            notification_data = data
+        elif not tx_sudo.acquirer_id.paypal_pdt_token:  # We received PDT data but can't verify them
+            raise ValidationError("PayPal: The PDT token is not set; cannot verify data origin")
+        else:  # The PayPal account is configured to receive PDT data, and the PDT token is set
+            # Request a PDT data authenticity check and the notification data to PayPal
+            url = tx_sudo.acquirer_id._paypal_get_api_url()
+            payload = {
+                'cmd': '_notify-synch',
+                'tx': data['tx'],
+                'at': tx_sudo.acquirer_id.paypal_pdt_token,
+            }
+            try:
+                response = requests.post(url, data=payload, timeout=10)
```

Comment:
Are you sure the information should be sent as form-encoded?

Reply:
Yes.